File: stacks/SortStack.java
package stacks;

import java.util.Stack;

public class SortStack {
    /*
        Sort Stack: Write a program to sort a stack such that the smallest items are on the top. You can use
        an additional temporary stack, but you may not copy the elements into any other data structure
        (such as an array). The stack supports the following operations: push, pop, peek, and is Empty
     */


    public static void sortStack(Stack<Integer> s1) {
        Stack<Integer> s2 = new Stack<>();

        sortHelper(s1, s2);
    }

    private static void sortHelper(Stack<Integer> s1, Stack<Integer> s2) {
        if (s1.isEmpty()) {
            return;
        }

        int temp = s1.pop();

        while (!s2.isEmpty() && s2.peek() > temp) {
            s1.push(s2.pop());
        }

        s2.push(temp);

        sortHelper(s1, s2);

        if (!s2.isEmpty()) {
            s1.push(s2.pop());
        }
    }

    public static void main(String[] args) {
        Stack<Integer> s1 = new Stack<>();
        s1.push(3);
        s1.push(5);
        s1.push(1);
        s1.push(4);
        s1.push(2);

        System.out.println("Original Stack: " + s1);
        sortStack(s1);
        System.out.println("Sorted Stack: " + s1);
    }

}

File: bin/stacks/SortStack.py
def sort_stack(s1):
    """
    Sorts a stack such that the smallest items are on the top.
    Uses a temporary stack (s2) and recursion.
    """
    s2 = []

    sort_helper(s1, s2)



def sort_helper(s1, s2 ):
    # Base case: if s1 is empty, return
    if not s1:
        return

    # Pop the top element from s1
    temp = s1.pop()

    # Move elements from s2 to s1 until the correct position for `temp` is found
    while s2 and s2[-1] > temp:
        s1.append(s2.pop())

    # Place the temp element in s2
    s2.append(temp)

    # Recur for the rest of s1
    sort_helper(s1, s2)

    # Move all sorted elements back from s2 to s1
    if s2:
        s1.append(s2.pop())


# Example usage
if __name__ == "__main__":
    s1 = [3, 5, 1, 4, 2]  # Python list as a stack
    print("Original Stack:", s1)
    sort_stack(s1)
    print("Sorted Stack:", s1)


